Highlight/Highlighter.php

```diff
@@ -104,17 +114,90 @@ public function __construct()
             ),
         );
 
-        self::registerLanguages();
+        if ($loadAllLanguages) {
+            self::registerAllLanguages();
+        }
     }
 
     /**
-     * Loop through all of the languages in our `languages` folder and automatically register them all.
+     * Return a list of all available languages bundled with this library.
+     *
+     * @since 9.18.1.4
+     *
+     * @return string[] An array of language names
+     */
+    public static function listBundledLanguages()
+    {
+        if (!empty(self::$bundledLanguages)) {
+            return self::$bundledLanguages;
+        }
+
+        // Languages that take precedence in the classMap array. (I don't know why...)
+        $bundledLanguages = array(
+            "xml" => true,
+            "django" => true,
+            "javascript" => true,
+            "matlab" => true,
+            "cpp" => true,
+        );
+
+        $languagePath = __DIR__ . '/languages/';
+        $d = @dir($languagePath);
+
+        if (!$d) {
+            throw new \RuntimeException('Could not read bundled language definition directory.');
+        }
+
+        // @TODO In 10.x, rewrite this as a generator yielding results
+        while (($entry = $d->read()) !== false) {
+            if (substr($entry, -5) === ".json") {
+                $languageId = substr($entry, 0, -5);
+                $filePath = $languagePath . $entry;
+
+                if (is_readable($filePath)) {
+                    $bundledLanguages[$languageId] = true;
+                }
+            }
+        }
+
+        $d->close();
+
+        return self::$bundledLanguages = array_keys($bundledLanguages);
+    }
+
+    /**
+     * Return a list of all the registered languages. Using this list in
+     * setAutodetectLanguages will turn on auto-detection for all supported
+     * languages.
+     *
+     * @since 9.18.1.4
      *
+     * @param bool $includeAliases Specify whether language aliases should be
+     *                             included as well
+     *
+     * @return string[] An array of language names
+     */
+    public static function listRegisteredLanguages($includeAliases = false)
+    {
+        if ($includeAliases === true) {
+            return array_merge(self::$languages, array_keys(self::$aliases));
+        }
+
+        return self::$languages;
+    }
+
+    /**
+     * Register all 185+ languages that are bundled in this library.
+     *
+     * To register languages individually, use `registerLanguage`.
+     *
+     * @since 9.18.1.4 Method is now public
      * @since 8.3.0.0
+     * @see Highlighter::registerLanguage
      *
      * @return void
      */
-    private static function registerLanguages()
+    public static function registerAllLanguages()
```

Comment:
I just realized that the static methods here may be problematic, no? If you have multiple instances of the class and each instance desires to have a different set of languages registered, is not the static `$languages` member variable problematic?

Reply:
Since the highlight.php 9.0.0 release, language registration has always been singleton-esque. So, having multiple instances of Highlighter has not given you a different selection of languages in recent times.

Whether or not this design will change in highlight.php 10 is TBD.